### PR TITLE
[SPARK-40324][SQL] Provide query context in AnalysisException

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

This is a major PR for extending AnalysisException with query context. After changes, we can:
* migrate analysis errors to error classes
* change the actual error messages of AnalysisException, showing the query context from original query instead of an unresolved logical plan.

### Why are the changes needed?

Make it possible for better error messages in analysis exceptions.

### Does this PR introduce _any_ user-facing change?

No. This is an extension of the framework. We will have more improvements based on this one.

### How was this patch tested?

Existing UT

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @srielau @anchovYu as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala [hunk 1]
```diff
@@ -124,12 +126,16 @@ class AnalysisException protected[sql] (
...
     new AnalysisException(message, line, startPosition, plan, cause, errorClass, errorSubClass,
-      messageParameters)
+      messageParameters, context)
 
-  def withPosition(line: Option[Int], startPosition: Option[Int]): AnalysisException = {
-    val newException = this.copy(line = line, startPosition = startPosition)
+  def withPosition(origin: Origin): AnalysisException = {
+    val newException = this.copy(
+      line = origin.line,
+      startPosition = origin.startPosition,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:137]: do we still need the `line` and `startPosition` parameters?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ExpressionTypeCheckingSuite.scala [hunk 2]
```diff
@@ -77,12 +77,6 @@ class ExpressionTypeCheckingSuite extends SparkFunSuite with SQLHelper {
   }
 
   test("check types for binary arithmetic") {
-    // We will cast String to Double for binary arithmetic
-    assertSuccess(Add($"intField", $"stringField"))
-    assertSuccess(Subtract($"intField", $"stringField"))
-    assertSuccess(Multiply($"intField", $"stringField"))
-    assertSuccess(Divide($"intField", $"stringField"))
-    assertSuccess(Remainder($"intField", $"stringField"))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ExpressionTypeCheckingSuite.scala:85]: why remove these?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:137]: oh we still need them, for errors without error classes.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ExpressionTypeCheckingSuite.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ExpressionTypeCheckingSuite.scala:85]: Removed by mistake. It is recovered now.

**Comment by dtenedor** [comment]
cc @dtenedor myself for context

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @gengliangwang and @cloud-fan for review.

**Pull request closed without merging**